### PR TITLE
feature: raw_timestamps support for writer and defragment

### PR DESCRIPTION
**Pull request by jimkring**
This PR adds support for writing raw_timestamps (thus preserving timestamp precision) when defragmenting a TDMS file.  This ensures that timestamp precision is preserved when defragmenting a TDMS file created with LabVIEW, for example.

Notable changes:

- `TdmsTimestamp` type now behave like `TdmsType` when writing bytes to file
  - added an `enum_value=0x44` class attribute
  - added a `bytes()` property that returns the struct.pack'ed data `bytes=_struct_pack('<Qq', second_fractions, seconds)`
- `nptdms.writer` now handles the `TdmsTimestamp` (in the `_to_tdms_value()` function)

### CODE REVIEW

**Comment by jimkring** [comment]
resolves #319

**Review by adamreeve — CHANGES_REQUESTED** [review]
Thanks for making a PR for this @jimkring! The approach looks good but I've got a few suggestions

File: nptdms/test/test_example_files.py [hunk 1]
```diff
@@ -33,6 +34,13 @@ def test_raw_format():
                                     0.2517777])
 
 
+def test_defragment_raw_timestamps():
+    """Test defragmenting a file with raw timestamps"""
+    test_file_path = DATA_DIR + '/raw_timestamps.tdms'
+    output_file = BytesIO()
+    TdmsWriter.defragment(test_file_path, output_file, raw_timestamps=True)
```
[the remark this answers is not in the document]
Comment (jimkring) [reply nptdms/test/test_example_files.py]: @adamreeve great idea. I've added assertions that the `channel.data` and `channel.time_track()` are the same in both the input and output data (using `np.testing.assert_equal`). I'm not sure if this is checking for loss of precision of the timestamps and could use your input on that.

File: nptdms/writer.py [hunk 2]
```diff
@@ -21,7 +22,7 @@ class TdmsWriter(object):
     """
 
     @classmethod
-    def defragment(cls, source, destination, version=4712, index_file=False):
+    def defragment(cls, source, destination, version=4712, index_file=False, raw_timestamps=False):
```
[the remark this answers is not in the document]
Comment (jimkring) [reply nptdms/writer.py]: I've made these changes. Feel free to review and add new comments. I'll close this comment.

File: nptdms/timestamp.py [hunk 3]
```diff
@@ -18,6 +20,8 @@ class TdmsTimestamp(object):
     def __init__(self, seconds, second_fractions):
         self.seconds = seconds
         self.second_fractions = second_fractions
+        self.enum_value = 0x44
+        self.bytes = _struct_pack('<Qq', second_fractions, seconds)
```
[the remark this answers is not in the document]
Comment (jimkring) [reply nptdms/timestamp.py]: done!

File: nptdms/test/test_example_files.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (adamreeve) [reply nptdms/test/test_example_files.py]: Yeah I don't think that's enough as the time track will be cast down to a numpy datetime64. To check that this test is actually testing the fix, it should fail if we remove `raw_timestamps=True` from `TdmsWriter.defragment`.

I think what you'll want to do is pass `raw_timestamps=True` when opening both `output_tdms` and `input_tdms`, and then add a check like:
```
# verify the written timestamp matches the input
output_channel = output_tdms[group.name][channel.name]
assert output_channel.properties['wf_start_time'] == channel.properties['wf_start_time']
```

To make that work correctly you'll also need to implement `__eq__` for `TdmsTimestamp`

[the remark this answers is not in the document]
Comment (jimkring) [reply nptdms/test/test_example_files.py]: @adamreeve BTW, I also just refactored the test_defragment to test more example files. In doing so, I noticed that one of the example files raises an error, due to scalar channel data.  However, this error should probably be addressed in a separate issue and PR.

File: nptdms/test/test_example_files.py [hunk 4]
```diff
@@ -33,6 +35,47 @@ def test_raw_format():
...
+            # verify channel data
+            np.testing.assert_equal(channel.data, tdms2[group.name][channel.name].data)
+            # verify channel properties
+            for p in channel.properties:
+                np.testing.assert_equal(channel.properties[p], tdms2[group.name][channel.name].properties[p])
+
+
+@pytest.mark.parametrize("tdms_file", [
+    'raw_timestamps.tdms',  # Test defragmentation of a file with raw timestamps
+    # 'raw1.tdms', # <- cannot defragment this file (ValueError: Channel data must be a 1d array)
```
Comment (jimkring) [inline nptdms/test/test_example_files.py:59]: @adamreeve note that trying to defragment this file fails due to it having scalar channel.data

Comment (adamreeve) [reply nptdms/test/test_example_files.py:59]: Right yes we don't yet support writing or defragmenting files with DaqMxRawData

**Review by adamreeve — APPROVED** [review]

**Comment by adamreeve** [comment]
This looks good now, thanks for fixing this @jimkring

**Comment by jimkring** [comment]
Thanks @adamreeve. Some of my teammates were interested in using the new version, once it's ready. Will there be a 1.7.2 release?

**Comment by adamreeve** [comment]
Sure, I've released [version 1.8.0](https://github.com/adamreeve/npTDMS/releases/tag/1.8.0) with these changes (new features should increment the minor version according to semver)

**Pull request merged**